server/EXCEPTIONS.md

```diff
@@ -0,0 +1,192 @@
+```
+
+Where...
+  - `status` is one of `HttpStatus` enum codes. E.g. `HttpStatus.UNAUTHORIZED`
+  - `responseBody` argument is an object with the following schema.
+    ```js
+       const responseBody = {
+        code: 'APPLICANT_PORTAL_ERROR_CODE',
+        message: 'Applicant Portal could not find coffee. CRM: CRM ran out of tea.',
+      }
+    ```
+    - The `code` property is developer-defined, and should be a unique, concise,
+    descriptive identifier for the type of error. Users can provide this code to
+    help developers locate the code that yielded the error.
+    - The `message` property is a short, human-readable sentence firstly
+    describing the error for the benefit of end users. Secondly, it should contain
+    errors relayed from CRM, if available.
+
+To scope the error to a specific part of the code, use try-catch blocks. For example,
+the following try-catch scopes the thrown error to an attempt to find a project.
+
+```ts
+// services/my-project.service.ts
+public async getProject(){
+  try {
+    return myProject = this.get('dcp_project', ...);
+  } catch (e) {
+    throw new HttpException({
+      code: 'GET_PROJECTS',
+      message: `Failed to find a project for given id. CRM: ${e.message}`,
+    });
+  }
+}
+```
+
+The `catch` block in a try-catch will catch uncaught errors generated from the `try` block, 
+
+Generally, when generating errors as described above, <u>*do so within Nest services*</u>.<br>
+Services are usually quite close to the actual CRM requests and contain
+specific backend business logic. We can provide better error details if we
+try to scope try-catch to more specifc areas of business logic.
+
+## Relaying Errors
+
+In the controller that consumes a service, _relay_ the error from the service by catching and re-throwing the error as it bubbles up:
+
+```ts
+// controllers/my-project.controller.ts
+@Get('/')
+async getProject(@Session() session) {
+  try {
+    return await this.myProjectService.getProject();
+  } catch (e) {
+    throw e; // simply relay the bubbling error
+  }
+}
+```
+
+Note how it is not necessary to construct a new HttpException in the controller to relay the error from the service. Instead, we simply use `throw` to pass along the error `e` from the service.
+
+The `catch` block in a try-catch will catch uncaught errors generated *within* functions that are called in the `try` block. Errors not yet caught will *bubble* through the function call stack until caught.
+
+<u>Ultimately, we are relaying errors up to the Nest global HttpException filter.</u> By re-throwing or generating a new error in catch blocks, we keep the error in an "uncaught" state. The global filter catches any uncaught errors and standardizes how they are wrapped in a request before being sent to the frontend (more in a following section). If in any try-catch blocks caught an error but did nothing in the catch block (relay it or generate a new error), the filter would never detect an uncaught exception.
```

Comment:
Interesting! I think I've been doing it wrong then...

Reply:
Have we confirmed this is true? I have been confused by this a lot

Reply:
Yep, I've experimented and verified this